File: 135-candy/135-candy.py
class Solution:
    def triangle_sum(self, n: int) -> int:
        return (n * (n + 1)) // 2
    # use slopes!
    # O(n) time, O(1) space!
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        if n == 1:
            return 1
        ups = 0
        downs = 0
        prev_diff = 0
        new_diff = 0
        candies = 0
        for i in range(1, n):
            if ratings[i] > ratings[i-1]:
                new_diff = 1
            elif ratings[i] < ratings[i-1]:
                new_diff = -1
            else:
                new_diff = 0
            if ((prev_diff > 0) and (new_diff == 0)) or ((prev_diff < 0) and (new_diff >= 0)):
                candies += self.triangle_sum(ups) + self.triangle_sum(downs) + max(ups, downs)
                ups, downs = 0, 0
            if new_diff > 0:
                ups += 1
            elif new_diff < 0:
                downs += 1
            else:
                candies += 1
            prev_diff = new_diff
        
        candies += 1 + self.triangle_sum(ups) + self.triangle_sum(downs) + max(ups, downs)
        
        return candies

File: 135-candy/135-candy.cpp
class Solution {
public:
    int triangle_sum(int n) {
        return (n * (n + 1)) / 2;
    }
    int max(int a, int b) {
        if (a >= b) {
            return a;
        }
        return b;
    }
    // use slopes!
    // O(n) time, O(1) space!
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        if (n == 1) {
            return 1;
        }
        int ups = 0;
        int downs = 0;
        int prev_diff = 0;
        int new_diff = 0;
        int candies = 0;
        for (int i=1; i<n; i++) {
            if (ratings[i] > ratings[i-1]) {
                new_diff = 1;
            }
            else if (ratings[i] < ratings[i-1]) {
                new_diff = -1;
            }
            else {
                new_diff = 0;
            }
            if (((prev_diff > 0) & (new_diff == 0)) || ((prev_diff < 0) & (new_diff >= 0))) {
                candies += this->triangle_sum(ups) + this->triangle_sum(downs) + this->max(ups, downs);
                ups = 0;
                downs = 0;
            }
            if (new_diff > 0) {
                ups++;
            }
            else if (new_diff < 0) {
                downs++;
            }
            else {
                candies++;
            }
            prev_diff = new_diff;
        
        }
        candies += 1 + this->triangle_sum(ups) + this->triangle_sum(downs) + this->max(ups, downs);
        
        return candies;
    }
};